Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp

```diff
@@ -0,0 +1,209 @@
+/* Copyright 2020 Remi Lehe
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "WarpXAlgorithmSelection.H"
+#include "FiniteDifferenceSolver.H"
+#ifdef WARPX_DIM_RZ
+#   include "FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H"
+#else
+#   include "FiniteDifferenceAlgorithms/YeeAlgorithm.H"
+#   include "FiniteDifferenceAlgorithms/CKCAlgorithm.H"
+#   include "FiniteDifferenceAlgorithms/NodalAlgorithm.H"
+#endif
+#include <AMReX_Gpu.H>
+
+using namespace amrex;
+
+/**
+ * \brief Update the B field, over one timestep
+ */
+void FiniteDifferenceSolver::EvolveB (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Bfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Efield,
+    amrex::Real const dt ) {
+
+   // Select algorithm (The choice of algorithm is a runtime option,
+   // but we compile code for each algorithm, using templates)
+#ifdef WARPX_DIM_RZ
+    if (m_fdtd_algo == MaxwellSolverAlgo::Yee){
+
+        EvolveBCylindrical <CylindricalYeeAlgorithm> ( Bfield, Efield, dt );
+
+#else
+    if (m_do_nodal) {
+
+        EvolveBCartesian <NodalAlgorithm> ( Bfield, Efield, dt );
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::Yee) {
+
+        EvolveBCartesian <YeeAlgorithm> ( Bfield, Efield, dt );
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::CKC) {
+
+        EvolveBCartesian <CKCAlgorithm> ( Bfield, Efield, dt );
+
+#endif
+    } else {
+        amrex::Abort("Unknown algorithm");
+    }
+
+}
+
+
+#ifndef WARPX_DIM_RZ
+
+template<typename T_Algo>
+void FiniteDifferenceSolver::EvolveBCartesian (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Bfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Efield,
+    amrex::Real const dt ) {
+
+    // Loop through the grids, and over the tiles within each grid
+#ifdef _OPENMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    for ( MFIter mfi(*Bfield[0], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+
+        // Extract field data for this grid/tile
+        auto const& Bx = Bfield[0]->array(mfi);
+        auto const& By = Bfield[1]->array(mfi);
+        auto const& Bz = Bfield[2]->array(mfi);
+        auto const& Ex = Efield[0]->array(mfi);
+        auto const& Ey = Efield[1]->array(mfi);
+        auto const& Ez = Efield[2]->array(mfi);
+
+        // Extract stencil coefficients
+        Real const* AMREX_RESTRICT coefs_x = stencil_coefs_x.dataPtr();
```

Comment:
```suggestion
        Real const * const AMREX_RESTRICT coefs_x = stencil_coefs_x.dataPtr();
```
Here it may matter that the pointer itself is `const`, at least for safety but maybe also for performance (not sure about that)?